overlord/devicestate/handlers.go

```diff
@@ -19,18 +19,59 @@
 package devicestate
 
 import (
+	"fmt"
+	"os/exec"
 	"time"
 
 	"gopkg.in/tomb.v2"
 
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/snapstate"
 	"github.com/snapcore/snapd/overlord/state"
 )
 
+func (m *DeviceManager) doMarkPreseeded(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	snaps, err := snapstate.All(st)
+	if err != nil {
+		return err
+	}
+
+	if m.preseed {
+		// unmount all snaps
+		for _, snapSt := range snaps {
+			inf, err := snapSt.CurrentInfo()
+			if err != nil {
+				return err
+			}
+			logger.Debugf("unmounting snap %s at %s", inf.InstanceName(), inf.MountDir())
+			if _, err := exec.Command("umount", "-d", "-l", inf.MountDir()).CombinedOutput(); err != nil {
+				return err
+			}
+		}
+
+		// do not mark this task done as this makes it racy against taskrunner tear down (the next task
+		// could start). Let this task finish after snapd restart when preseed mode is off.
+		st.RequestRestart(state.StopSnapd)
+
+		return &state.Retry{Reason: "mark-preseeded will be marked done when snapd is executed in normal mode"}
```

Comment:
aren't we going to risk doing the above again and again if there's some extra EnsureBefore(0) ?

Reply:
Fair point. I think in theory it shouldn't happen because we have precise list of tasks at this point, but it's a good idea to have a safeguard. I've added a flag on the task.

Reply:
Also updated the test to make sure this task can be re-tried with preseeding=true without undesired effects.